DrawMat/Utilities/AvaloniaExtensions.cs

```diff
@@ -7,38 +7,36 @@ public static class RectExtensions
 {
     public static Rect Union(this Rect a, Rect b)
     {
+        if (a.IsEmpty()) return b;
+        if (b.IsEmpty()) return a;
+
         double x1 = Math.Min(a.X, b.X);
         double y1 = Math.Min(a.Y, b.Y);
-        double x2 = Math.Max(a.X + a.Width, b.X + b.Width);
-        double y2 = Math.Max(a.Y + a.Height, b.Y + b.Height);
+        double x2 = Math.Max(a.Right, b.Right);
+        double y2 = Math.Max(a.Bottom, b.Bottom);
 
         return new Rect(x1, y1, x2 - x1, y2 - y1);
     }
 
     public static Rect Union(this Rect rect, Point pt)
     {
-        double x1 = Math.Min(rect.X, pt.X);
-        double y1 = Math.Min(rect.Y, pt.Y);
-        double x2 = Math.Max(rect.Right, pt.X);
-        double y2 = Math.Max(rect.Bottom, pt.Y);
-
-        return new Rect(x1, y1, x2 - x1, y2 - y1);
+        return RectExtensions.Union(rect, new Rect(pt.X, pt.Y, 0, 0));
     }
 
     public static Rect Union(this Point pt1, Point pt2)
     {
-        double x1 = Math.Min(pt1.X, pt2.X);
-        double y1 = Math.Min(pt1.Y, pt2.Y);
-        double x2 = Math.Max(pt1.X, pt2.X);
-        double y2 = Math.Max(pt1.Y, pt2.Y);
-
-        return new Rect(x1, y1, x2 - x1, y2 - y1);
+        return RectExtensions.Union(new Rect(pt1.X, pt1.Y, 0, 0), new Rect(pt2.X, pt2.Y, 0, 0));
     }
 
     public static Rect AddMargin(this Rect rect, double Margin)
     {
         return new Rect(rect.X - Margin/2, rect.Y - Margin/2, rect.Width + Margin, rect.Height + Margin);
-    }    
+    }
+    
+    public static bool IsEmpty(this Rect rect)
+    {
+        return rect.X <= 0 && rect.Y <= 0 && rect.Width <= 0 && rect.Height == 0;
```

Comment:
`new Rect(x, y, 0, 0).Contains(new Point(x, y))` is true for all `x` and `y`, so `new Rect(x, y, 0, 0).IsEmpty().Should().BeFalse()` and the implementation of `IsEmpty` can only depend on `Width` and `Height`. Think about which values make `new Rect(x, y, width, height).Contains(new Point(x, y))` false, so that `new Rect(x, y, width, height).IsEmpty().Should().BeTrue()`.